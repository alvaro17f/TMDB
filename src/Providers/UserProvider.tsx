import { createContext, useContext, useReducer } from "react";
import { GuestActionType, GuestType, UserContextType } from "../types/types";

const UserContext = createContext<UserContextType>({});

const useUserContext = () => {
  return useContext(UserContext);
};

const initialState = {
  success: undefined,
  guest_session_id: undefined,
  expires_at: undefined,
};

const reducer = (state: GuestType, action: GuestActionType): GuestType => {
  switch (action.type) {
    case "CREATE_GUEST_SESSION": {
      return {
        ...state,
        success: action.payload.success,
        guest_session_id: action.payload.guest_session_id,
        expires_at: action.payload.expires_at,
      };
    }
    default: {
      return state;
    }
  }
};

const UserProvider = ({ children }: { children: React.ReactNode }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <UserContext.Provider value={{ state, dispatch }}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider, useUserContext };
