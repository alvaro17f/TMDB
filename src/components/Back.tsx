import { useNavigate } from "react-router-dom";

export const Back = () => {
  const navigate = useNavigate();
    return (
      <div className="grid h-16 border-b-2 place-content-start">
        <button
          type="button"
          className="w-24 h-16 text-xl text-slate-600"
          onClick={() => navigate(-1)}
        >
          Back
        </button>
      </div>
    );
};
