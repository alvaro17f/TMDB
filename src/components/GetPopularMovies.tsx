import { useEffect, useState } from "react";

// type RootObject = {
//   page: number;
//   results: Result[];
//   total_pages: number;
//   total_results: number;
// }

type Results = {
  adult: boolean;
  backdrop_path: string;
  genre_ids: number[];
  id: number;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string;
  release_date: Date;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}[];

const URL = `https://api.themoviedb.org/3/movie/popular?api_key=${
  import.meta.env.VITE_API_KEY
}`;
const IMG_URL = "https://image.tmdb.org/t/p/original";

export const GetPopularMovies = () => {
  const [popularMovies, setPopularMovies] = useState<Results>();

  useEffect(() => {
    fetch(URL)
      .then((resp) => resp.json())
      .then((movies) => setPopularMovies(movies.results));
  }, []);

  return (
    <section className="grid md:grid-cols-4 place-items-center gap-5 m-10">
      {popularMovies?.map((movie) => (
        <div id={String(movie.id)} className="border rounded-xl w-56 h-[30rem]">
          <img src={IMG_URL + movie.poster_path} alt={movie.title} />
          <div className="p-5 relative h-32">
            <h1 className="text-lg max-w-[200px] text-center">{movie.title}</h1>
            <p className="text-xs text-center text-slate-700 absolute bottom-0">{`Fecha de Lanzamiento: ${new Date(
              movie.release_date,
            ).toLocaleDateString()}`}</p>
          </div>
        </div>
      ))}
    </section>
  );
};
