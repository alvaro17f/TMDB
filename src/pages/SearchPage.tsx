import { Link } from "react-router-dom";
import { useFetchSearch } from "../hooks/useFetchSearch";
import { Paginator } from "../components/Paginator";

const IMG_URL = "https://image.tmdb.org/t/p/original";

export const SearchPage = () => {
  const { data, loading, results } = useFetchSearch();

  if (loading) return <div className="text-5xl text-center">Loading ...</div>;
  if (data.length === 0)
    return <div className="text-5xl text-center">No results</div>;
  return (
    <>
      <section className="grid gap-5 m-10 md:grid-cols-4 place-items-center">
        {data?.map((movie) => (
          <Link
            id={String(movie.id)}
            to={`/movie/${movie.id}`}
            className="border rounded-xl w-56 h-[30rem]"
          >
            <img src={IMG_URL + movie.poster_path} alt={movie.title} />
            <div className="relative h-32 p-5">
              <h1 className="text-lg max-w-[200px] text-center">
                {movie.title}
              </h1>
              <p className="absolute bottom-0 text-xs text-center text-slate-700">{`Release date: ${new Date(
                movie.release_date,
              ).toLocaleDateString()}`}</p>
            </div>
          </Link>
        ))}
      </section>
      <Paginator results={results} />
    </>
  );
};