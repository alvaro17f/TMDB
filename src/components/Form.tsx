import { ChangeEvent, FormEvent, useState } from "react";
import { useFetchRating } from "../hooks/useFetchRating";
import { useUserContext } from "../Providers/UserProvider";

export const Form = () => {
  const [rating, setRating] = useState(2.5);
  
  const { state } = useUserContext();

  const handleRating = (e: ChangeEvent<HTMLInputElement>) => {
    setRating(Number(e.target.value));
  };

  const { data, loading, request } = useFetchRating({ value: rating });
  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    request();
  };
  if (loading) return <div className="text-2xl text-center">LOADING...</div>;
  if (!state.guest_session_id) return <div className="text-2xl text-center">You must click on "ğŸ˜€ Get ID" first...</div>;
  return (
    <form onSubmit={handleSubmit}>
      <div className="grid grid-cols-2 p-5 place-items-center">
        <div>
          <input
            type="range"
            min={0.5}
            max={5}
            step={0.5}
            value={rating}
            onChange={handleRating}
          />
          <label htmlFor="rating" className="p-5">
            rating: {rating * 2}
          </label>
        </div>
        <button type="submit" className="max-w-lg p-5 rounded-lg bg-slate-400">
          Submit
        </button>
      </div>
      {data && (
        <div className="grid p-5 border border-slate-400 rounded-xl w-fit place-items-center">
          {data?.status_message}
        </div>
      )}
    </form>
  );
};
