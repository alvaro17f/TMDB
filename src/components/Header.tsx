import { Link } from "react-router-dom";
import { Search } from "./Search";
import { useFetchGuest } from "../hooks/useFetchGuest";
import { useUserContext } from "../Providers/UserProvider";

export const Header = () => {
  const { state } = useUserContext();
  const { request, error } = useFetchGuest();

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    request();
  };
  return (
    <header className="flex items-center h-20 pl-5 shadow-md">
      <div className="text-5xl text-slate-400">
        <Link to="/">TMDB</Link>
      </div>
      <div className="flex items-baseline gap-5 m-auto mr-5 text-gray-800">
        {state?.guest_session_id ? (
          <p>
            Expires at:{" "}
            {new Date(state.expires_at as Date).toLocaleDateString()}
          </p>
        ) : error ? (
          <button type="button" onClick={handleClick}>
            Error getting an ID
          </button>
        ) : (
          <button type="button" onClick={handleClick}>
            😀 Get ID
          </button>
        )}
        <Link to="/mylist">⭐ List</Link>
        <Search />
      </div>
    </header>
  );
};
