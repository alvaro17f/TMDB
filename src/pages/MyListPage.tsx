import { Link } from "react-router-dom";
import { useFetchList } from "../hooks/useFetchList";
import { useUserContext } from "../Providers/UserProvider";
import { useFetchGuest } from "../hooks/useFetchGuest";

const IMG_URL = "https://image.tmdb.org/t/p/original";

export const MyListPage = () => {
  const { data, loading } = useFetchList();
  const { request } = useFetchGuest();
  const { state } = useUserContext();

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    request();
  };

  if (loading)
    return <div className="mt-5 text-5xl text-center">Loading ...</div>;
  if (!state?.guest_session_id)
    return (
      <div className="mt-5 text-2xl text-center ">
        You must click on{" "}
        <button
          type="button"
          className="p-2 border rounded-xl border-slate-400 bg-slate-300"
          onClick={handleClick}
          aria-label="get-id-list"
        >
          ðŸ˜€ Get ID{" "}
        </button>{" "}
        to see your list...
      </div>
    );
  return (
    <section className="m-5">
      <h1 className="mb-5 text-5xl">MY LIST</h1>
      <table className="w-full border ">
        <thead className="h-10 text-left bg-slate-400">
          <tr>
            <th />
            <th>Title</th>
            <th>Vote Average</th>
            <th>Personal Rating</th>
          </tr>
        </thead>
        <tbody>
          {data?.map(({ rating, title, poster_path, vote_average, id }) => (
            <tr key={id} className="border">
              <td className="w-36">
                <Link to={`/movie/${id}`}>
                  <img
                    src={IMG_URL + poster_path}
                    alt={title}
                    className="w-32"
                  />
                </Link>
              </td>
              <td>
                <Link to={`/movie/${id}`}>{title}</Link>
              </td>
              <td>{vote_average.toPrecision(2)}</td>
              <td>{rating * 2}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </section>
  );
};
