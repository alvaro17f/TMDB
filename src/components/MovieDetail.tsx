import { useFetchMovies } from "../hooks/useFetchMovies";
import { Form } from "./Form";

const IMG_URL = "https://image.tmdb.org/t/p/original";

export const MovieDetail = ({ id }: { id: string }) => {
  const { data: movie, loading } = useFetchMovies(String(id));

  return loading ? (
    <div className="grid text-5xl place-content-center">Loading...</div>
  ) : (
    <div className="grid grid-cols-[1fr_2fr]">
      <div>
        <img
          src={IMG_URL + movie?.poster_path}
          alt={movie?.title}
          className="mb-5 w-96"
        />
      </div>
      <div className="grid">
        <h1 className="text-5xl">{movie?.title}</h1>
        <p className="absolute right-0 mr-12 text-xs">ID: {movie?.id}</p>
        <p>{`Release date: ${new Date(
          movie?.release_date as Date,
        ).toLocaleDateString()}`}</p>
        <p>{movie?.adult ? "Adults Only" : "Every Ages"}</p>
        <p>{`Votes: ${movie?.vote_count}`}</p>
        <p>{movie?.overview}</p>
        <Form />
      </div>
    </div>
  );
};
