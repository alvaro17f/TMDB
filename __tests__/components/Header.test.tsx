import React from "react";
import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { Header } from "../../src/components/Header";
import { MyListPage } from "../../src/pages/MyListPage";
import { SearchPage } from "../../src/pages/SearchPage";
import { Routes, Route, MemoryRouter } from "react-router-dom";
import { UserProvider } from "../../src/Providers/UserProvider";
import { server } from "../../src/mocks/server";
import { rest } from "msw";

const wrapper = (
  <UserProvider>
    <MemoryRouter>
      <Routes>
        <Route path="/" Component={Header} />
        <Route path="/mylist" Component={MyListPage} />
        <Route path="/search/:searchQuery/:page" Component={SearchPage} />
      </Routes>
    </MemoryRouter>
  </UserProvider>
);
describe("<Header />", () => {
  beforeEach(() => {
    render(wrapper);
    afterEach(cleanup);
  });
  it("should match the snapshot", () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

  it("should render LOGO", () => {
    screen.getByText(/tmdb/i);
  });

  it("should render GET ID button", () => {
    screen.getByText(/get id/i);
  });

  it("should render 'Expires at...' after click", async () => {
    fireEvent.click(screen.getByRole("button", { name: /get id/i }));
    fireEvent.click(screen.getByText(/get id/i));

    await waitFor(() => {
      screen.getByText(/expires at: /i);
    });
  });

  it("should render List button", () => {
    screen.getByText(/list/i);
  });

  it("should navigate to /mylist after clicking on List button", async () => {
    fireEvent.click(screen.getByText(/list/i));
    await waitFor(() => {
      screen.getByText(/get id/i);
    });
  });

  it("should render search component", async () => {
    screen.getByPlaceholderText(/john wick/i);
    screen.getByRole("button", { name: /search/i });
    screen.getByRole("searchbox");
    fireEvent.click(screen.getByText(/search/i));
    fireEvent.change(screen.getByRole("searchbox"), {
      target: { value: "break" },
    });
    fireEvent.click(screen.getByText(/search/i));
    screen.getByText(/loading.../i);
    await waitFor(() => {
      screen.getAllByText(/break/i);
    });
    fireEvent.click(screen.getByText(">"));
  });
});
