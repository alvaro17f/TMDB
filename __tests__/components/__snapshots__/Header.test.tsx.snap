// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> > should match the snapshot 1`] = `
<div>
  <header
    class="flex items-center h-20 pl-5 shadow-md"
  >
    <div
      class="text-5xl text-slate-400"
    >
      <a
        href="/"
      >
        TMDB
      </a>
    </div>
    <div
      class="flex items-baseline gap-5 m-auto mr-5 text-gray-800"
    >
      <button
        type="button"
      >
        😀 Get ID
      </button>
      <a
        href="/mylist"
      >
        ⭐ List
      </a>
      <form
        class="flex flex-row-reverse"
      >
        <input
          class="flex-1 p-2 border rounded-xl focus:outline-none"
          placeholder="John Wick"
          type="search"
          value=""
        />
        <button
          class="text-white absolute p-[9px] bg-slate-400 rounded-xl outline-none cursor-default"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </header>
</div>
`;
